pkg/handlers/handlers.go

```diff
@@ -303,6 +311,19 @@ func HandleFile(
 		return nil
 	}
 
+	// If the file is an 'application/jar' or "application/zip" file, check filename for the .apk extension.
```

Comment:
suggestion (optional): I’d suggest moving all this logic into `newFileReader` to keep `HandleFile` as simple and streamlined as possible.

Have you considered optionally passing a second argument to `newFileReader` containing the file extension? This could serve as a hint to supplement the processing logic.

So for example:
```
type readerConfig struct{ fileExtension string }

type readerOption func(*readerConfig)

func withFileExtension(ext string) readerOption {
	return func(c *readerConfig) { c.fileExtension = ext }
}

// newFileReader creates a fileReader from an io.Reader, optionally using BufferedFileWriter for certain formats.
func newFileReader(r io.Reader, options ...readerOption) (fileReader, error) {
	var (
		fReader fileReader
		cfg     readerConfig
	)

	for _, opt := range options {
		opt(&cfg)
	}

	fReader.BufferedReadSeeker = iobuf.NewBufferedReaderSeeker(r)

	mime, err := mimetype.DetectReader(fReader)
	if err != nil {
		return fReader, fmt.Errorf("unable to detect MIME type: %w", err)
	}
	fReader.mime = mimeType(mime.String())
	fReader.ext = mime.Extension()

	// Reset the reader to the beginning because DetectReader consumes the reader.
	if _, err := fReader.Seek(0, io.SeekStart); err != nil {
		return fReader, fmt.Errorf("error resetting reader after MIME detection: %w", err)
	}

	const apkExt = ".apk"
	// Check for APK files.
	if cfg.fileExtension == apkExt && (fReader.mime == zipMime || fReader.mime == jarMime || fReader.mime == warMime) {
		isAPK, err := isAPKFile(&fReader)
		if err != nil {
			return fReader, fmt.Errorf("error checking for APK: %w", err)
		}
		if isAPK {
			fReader.mime = apkMime
			fReader.ext = apkExt
			if _, err := fReader.Seek(0, io.SeekStart); err != nil {
				return fReader, fmt.Errorf("error resetting reader after APK detection: %w", err)
			}
			return fReader, nil
		}
	}

	// If a MIME type is known to not be an archive type, we might as well return here rather than
	// paying the I/O penalty of an archiver.Identify() call that won't identify anything.
	if _, ok := skipArchiverMimeTypes[mimeType(mime.String())]; ok {
		return fReader, nil
	}

	format, _, err := archiver.Identify("", fReader)
	switch {
	case err == nil:
		fReader.isGenericArchive = true
		fReader.format = format

	case errors.Is(err, archiver.ErrNoMatch):
		// Not an archive handled by archiver.
		// Continue with the default reader.
	default:
		return fReader, fmt.Errorf("error identifying archive: %w", err)
	}

	// Reset the reader to the beginning again to allow the handler to read from the start.
	// This is necessary because Identify consumes the reader.
	if _, err := fReader.Seek(0, io.SeekStart); err != nil {
		return fReader, fmt.Errorf("error resetting reader after archive identification: %w", err)
	}

	return fReader, nil
}

func isAPKFile(r *fileReader) (bool, error) {
	size, _ := r.Size()
	zipReader, err := zip.NewReader(r, size)
	if err != nil {
		return false, fmt.Errorf("error creating zip reader: %w", err)
	}

	hasManifest := false
	hasClasses := false

	for _, file := range zipReader.File {
		switch file.Name {
		case "AndroidManifest.xml":
			hasManifest = true
		case "classes.dex":
			hasClasses = true
		default:
			// Skip other files.
		}
		if hasManifest && hasClasses {
			return true, nil
		}
	}

	return false, nil
}
```

I’d also argue that including the slightly expensive `isAPKFile` check in `newFileReader` is worth it. This avoids falling back to handle a partially consumed reader, making the code easier to reason about. Ensuring that the reader returned by `newFileReader` already includes the necessary information for the strategy pattern eliminates the need for a fallback and cleans up the main `HandleFile` call site.

I also think the cost of calling `getFileExtension` on each file seems acceptable. While it would be ideal to extract the file extension directly from metadata for clearnly, that’s a separate issue to address later

Reply:
I like this approach. If we eliminate the fallback `archive/zip` processing in `apk.HandleFile()`, are we fine with the edge case where there is a `.apk` file containing `AndroidManifest.xml` and `classes.dex` that is actually a non-apk archive that happens to contain those two files? Seems like an extreme edge case, but just wanted to check.

Reply:
I’d vote yes. Avoiding support for such an edge case simplifies the code, which feels like a worthwhile trade-off. We can always leave a TODO and comment to revisit if it becomes a significant issue later. That said, I’m definitely open to hearing opposing thoughts, I might just be too close to this code to give a completely un-biased opinion. 😅